library/network_connections.py

```diff
@@ -465,7 +465,12 @@ def ifcfg_create(
                     line += " via " + r["gateway"]
                 if r["metric"] != -1:
                     line += " metric " + str(r["metric"])
-
+                if r["dev"]:
+                    line += " dev " + str(r["dev"])
```

Comment:
Oh no, I did not know that. My tests prove that having ".... dev eth1" in rule-eth1 still does not break anything, but I agree it makes no sense to use it in this role. I will remove it.

Reply:
@thom311 you are not completely right. In `ifcfg-routes` of my Centos 7.6 installation there are two functions: `handle_file()`, which contains those lines you mention, and `handle_ip_file` which does NOT add `dev` as in `handle_file()`. So removing this parameter from my routing configuration effectively disabled the config, because the second line from the `ifcfg-eth0` looks like this:

```
# cat route-eth0
0.0.0.0/0 via 11.224.79.49 table 1
11.224.79.48/28 scope link table 1
```

and executing 

```
# ip ro add 11.224.79.48/28 scope link table 1
```

results in:
```
RTNETLINK answers: No such device
```

Reply:
you are right. Still "dev" should not be something that the user configures in the playbook, but the role should figure it out itself. Like it does for 
```
            ethtool_options += "-K %s %s" % (
                connection["interface_name"],
                " ".join(configured_features),
            )
```

Reply:
Hm, that's a hard job, because we need to decide which routing rule requires `dev` and which not. For example, dev is invalid for blackhole routes:

```
# ip ro add blackhole 0.0.0.0/0 table 3 dev eth0
RTNETLINK answers: Invalid argument
```

I am not sure we should implement the logic of attribute combo validation into the role :/

Reply:
blackhole routes are currently not supported by the role.
And if support is added, then it's simple for the role to know whether to add `dev` or not.

The entire point is that the role fully understands what it is configuring there. It's not merely passing some command line parameters through to ip-route.

Reply:
It is implemented, on my laptop :)

The point is, I know it's not needed by blackhole roles but I have no god damn idea which routes expect dev and which not. But okay, let me dive into it...

Reply:
> which routes expect dev and which not.

You support router of a certain type. For these types you know it. The ones where you don't know it don't matter, because they are not supported.